OpenBlas: add constraint for required versions of CMake when compiling with MSVC

johnwparent opened the pull request:
CMake 4.1 adds https://gitlab.kitware.com/cmake/cmake/-/merge_requests/10836 which is required to build openblas with MSVC + oneAPI

Presumably more packages will have this issue, it's not an issue with openblas' CMake, but openblas is the package that allowed us to create the fix for CMake, so it gets the constraint first as its the only one I'm currently aware of

mathomp4 reviewed (approved):
Trust in @johnwparent as I have no idea how to test this :) Approve

johnwparent commented:
> Trust in @johnwparent as I have no idea how to test this :) Approve

As much as I appreciate that, if you did want to test:

On Spack Develop on Windows:
- build this package using oneAPI and MSVC using CMakes before 4.1 and 4.1, and verify it works for 4.1 (something like `spack install "openblas ^^cmake@4.0"` and then `spack install "openblas ^^cmake@4.1"`)

It should fail on the first and pass on the second. Basically we found a bug in CMake trying to build this on Windows, so the real "test" is actually in the CMake codebase (see linked CMake issue)

If you're unable to get your hands on a Windows machine, I can provide a number of suggestions on how to do that, from spinning up a VM to SSH-ing into a GHA runner.

johnwparent commented:
@spackbot fix style

johnwparent committed:
CMake: Version constraint when using OneAPI, MSVC

CMake pre 4.1 does not alwyas properly handle linker argument forwarding
when mixing MSVC and OneAPI.

Add a conflict to the CMakePackage class that prevents useage of CMake
per 4.1 when using msvc as a cxx compiler and oneapi as a fortran compiler

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
CMake: further constraint <4.1 conflict

Ensure conflict is only applied when msvc provides
both fortran and cxx

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent commented:
@spackbot fix style

johnwparent committed:
[@spackbot] updating style on behalf of johnwparent
